File: task.py
'''
Created on 05.08.2020

@author: Max Weise
'''
from time import time  # Used for the stoppwatch
from datetime import timedelta  # Convert seconds to format h:mm:ss
import CustomExceptions as cExcept

''' This file contains the Task class. The executable program is written in the 'main.py' file.

#=== -> is used to get rid of indentation guide (closing bracket, so to speak)
'''

# NOTE: Not shure if id for each task is necessary / good idea 


class Task(object):
    ''' Blueprint for a task
    @author: Max Weise

    name : str                -> descriptive name of the task
    module : char             -> code, plan, debug
    time : int                -> time spent on given task
    work_in_progress : bool   -> is the task in progress?
    priority : int [-1, 0, 1] -> 0 = normal, the higher the number, the higher the priority
    '''

    def __init__(self, name, module, priority=0, time=0, work_in_progress=False):
        self.name = ''
        self.changeName(name)

        self.module = module
        self.time = time
        self.wip = work_in_progress
        self.priority = priority
#         self.unique_identification_number = id(self)

    def __str__(self):
        # getting name of the module
        if self.module == 'c':
            m = 'code'
        elif self.module == 'd':
            m = 'debug'
        else:
            m = 'plan'

        return f'[{self.priority}] \'{self.name}\' ({m} | T: {timedelta(seconds=self.time)})'
#        return f'[{self.priority}] \'{self.name}\' ({m} | T: {timedelta(seconds=self.time)}) <{self.unique_identification_number}>'

    def toDict(self):
        '''Returns the object as a dictionary'''
        list_of_keys = ['name', 'module', 'time', 'priority', 'work in progress']
        list_of_values = [self.name, self.module, self.time, self.priority, self.wip]

        taskDict = dict(zip(list_of_keys, list_of_values))
        return taskDict

    def changeName(self, name_to_change):
        '''Change name of task. Name may not contain spaces'''
        invalid_name = name_to_change  # To check if name has been modified

        try:
            if name_to_change.__contains__(' '):
                raise cExcept.NameException

        except Exception as e:
            print(e.message)
            name_to_change = name_to_change.replace(' ', '_')

        finally:
            if invalid_name != name_to_change:
                print(f'\'{invalid_name}\' was invalid, new name is : \'{name_to_change}\'')

            self.name = name_to_change

    def stoppwatch(self):
        ''' Get the time used for a specific task
        @author: Max Weise
        returns time in seconds
        Use 'timedelta' to get the h:mm:ss format
        '''
        input('Press enter to start the stopwatch')
        print('Started stopwatch')

        t0 = time()

        input('To stop, press enter')

        t1 = time()
        delta_t = t1 - t0

        delta_t = int(delta_t)

        self.time += delta_t
        print(f'Stopwatch: {timedelta(seconds=delta_t)} | Total: {timedelta(seconds=self.time)}')


if __name__ == '__main__':
    print('This module is used for declarations only.\nTo run the program, please use the \'main.py\' file')

File: Task.java
package organisationsprogramm_v2_java;

import java.util.Scanner;

// import org.apache.commons.lang.time.StopWatch;

/**
 * @author Max Weise
 * 
 * Created on 04.09.2020
 * Class to model all the given tasks in the programm.
 */
class Task {

	/**
	 * name     -> descriptive name of the task
	 * module   -> code, plan, debug
	 * time     -> time in seconds, total time spent on task
	 * wip      -> is task in progress?
	 * priority -> -1, 0, 1; 0 is normal, the higher, the more important
	 */

	String name;
	char module;
	int time;
	boolean work_in_progress;
	int priority;

	public Task(String name, char module, int time, boolean wip, int priority) {
		this.name = this.setName(name);
		this.module = module;
		this.time = time;
		this.work_in_progress = wip;
		this.priority = priority;
	}

	/**
	 * Sets name for task. (May not contain spaces, but has yet to be implemented)
	 * @param nameToChange (Will later be processed for spaces)
	 * @return nameToChange
	 */
	String setName(String nameToChange) {
		return nameToChange;
	}

	/**
	 * Used to measure time in milliseconds, ms gets converted into seconds. 
	 * @return time passed in seconds
	 */
	int stopwatch() {
		int seconds = 0;
		long t0 = 0;
		long t1 = 0;
		
		System.out.println("Started stopwatch");
		
		t0 = System.currentTimeMillis();
		
		System.out.println("Type <end> to stop the stopwatch");
		Scanner endStopwatch = new Scanner(System.in);
		String a = endStopwatch.nextLine();
		
		if(a == "")
		
		endStopwatch.close();
		t1 = System.currentTimeMillis();
			
		seconds = (int) (t1 - t0) / 1000;
		this.time += seconds;
		
		return seconds;
	}

	/**
	 * overrides build in toString method 
	 * @return Task in following format: [priority] name (module | time )
	 */
	@Override
	public String toString() {
		return "[" + this.priority + "] " + this.name + " (" + this.module + "| T: " + this.time + " )";

	}
} // End of Class